source/4_tools/regina/src/regina.cpp

```diff
@@ -1,54 +1,865 @@
+
+                renderer::ResourceSlot curr_object_cr = get_active_object_commited_resource_for_frame(renderer::active_frame()->slot_id());
+                renderer::update_commited_resource(rsl::move(update_constant_buffer_command_desc), curr_object_cr);
+
+                // Updating constant buffer of the cube so the frame should remain dirty
+                ri.num_frames_dirty = renderer::num_frames_in_flight();
+            }
+        }
+    }
+    //-------------------------------------------------------------------------
+    void update_pass_constant_buffers()
+    {
+        renderer::ResourceSlot curr_pass_cr = get_active_pass_commited_resource_for_frame(renderer::active_frame()->slot_id());
+
+        const glm::mat4& view = g_regina_ctx.view;
+        const glm::mat4& proj = g_regina_ctx.proj;
+
+        glm::mat4 view_proj = view * proj;
+
+        glm::mat4 inv_view = glm::inverse(view);
+        glm::mat4 inv_proj = glm::inverse(proj);
+        glm::mat4 inv_view_proj = glm::inverse(view_proj);
+
+        g_regina_ctx.pass_constants.view = view;
+        g_regina_ctx.pass_constants.inv_view = inv_view;
+        g_regina_ctx.pass_constants.proj = proj;
+        g_regina_ctx.pass_constants.inv_proj = inv_proj;
+        g_regina_ctx.pass_constants.view_proj = view_proj;
+        g_regina_ctx.pass_constants.inv_view_proj = inv_view_proj;
+
+        f32 window_width = static_cast<f32>(globals::window_info().width);
+        f32 window_height = static_cast<f32>(globals::window_info().height);
+
+        g_regina_ctx.pass_constants.eye_pos_w = g_regina_ctx.eye_pos;
+        g_regina_ctx.pass_constants.render_target_size = glm::vec2(window_width, window_height);
+        g_regina_ctx.pass_constants.inv_render_target_size = glm::vec2(1.0f / window_width, 1.0f / window_height);
+        g_regina_ctx.pass_constants.near_z = globals::default_depth_info().near_plane;
+        g_regina_ctx.pass_constants.far_z = globals::default_depth_info().far_plane;
+        g_regina_ctx.pass_constants.delta_time = globals::frame_info().delta_time().to_seconds();
+
+        renderer::commands::UpdateCommitedResourceCommandDesc update_constant_buffer_command_desc;
+        update_constant_buffer_command_desc.element_index = 0;
+        update_constant_buffer_command_desc.buffer_data = memory::make_blob((rsl::byte*)&g_regina_ctx.pass_constants, rsl::memory_size(sizeof(PassConstants)));
+
+        renderer::update_commited_resource(rsl::move(update_constant_buffer_command_desc), curr_pass_cr);
+    }
+
+    //-------------------------------------------------------------------------
+    bool initialize()
+    {
+        REX_LOG(LogRegina, "Initializing Regina");
+
+        g_regina_ctx.scene = rsl::make_unique<renderer::Scene>();
+        g_regina_ctx.scene_renderer = rsl::make_unique<renderer::SceneRenderer>(g_regina_ctx.scene.get());
+
+        if (!build_clear_state()) return false;
```

Comment:
I think these graphics initializations should moved into its own function so the initialize function of regina only gives a high level overview of what gets initialized

Reply:
I don't understand the question here. the graphics initializations are within there own functions already?

The clear state should also be managed by the editor, maybe a different clear color has to be assigned, maybe nothing needs to be cleared, a different depth value has to be assigned. A stencil value needs to be assigned. Multiple use cases come to mind why you would want the user to give control over this. That said a default one can be registered by the engine if nothing was specified.

Reply:
I added the extra abstraction